code/game/machinery/firealarm.dm

```diff
@@ -306,104 +307,180 @@
-						span_notice("You remove the fire alarm assembly from the wall."))
-					var/obj/item/wallframe/firealarm/frame = new /obj/item/wallframe/firealarm()
-					frame.forceMove(user.drop_location())
-					tool.play_tool_sound(src)
-					qdel(src)
-					return
-	return ..()
+	loc.balloon_alert_to_viewers("removing circuit...")
+	if(!tool.use_tool(src, user, 2 SECONDS, volume = 50, extra_checks = CALLBACK(src, PROC_REF(state_callback), FIRE_ALARM_BUILD_NO_WIRES, TRUE)))
+		return ITEM_INTERACT_BLOCKING
+	if(machine_stat & BROKEN)
+		balloon_alert(user, "broken circuit removed")
+		set_machine_stat(machine_stat & ~BROKEN)
+	else
+		balloon_alert(user, "circuit removed")
+		new /obj/item/electronics/firealarm(user.drop_location())
+	buildstage = FIRE_ALARM_BUILD_NO_CIRCUIT
+	update_appearance()
+	return ITEM_INTERACT_SUCCESS
+
+/obj/machinery/firealarm/crowbar_act_secondary(mob/living/user, obj/item/tool)
+	return crowbar_act(user, tool)
+
+/obj/machinery/firealarm/wrench_act(mob/living/user, obj/item/tool)
+	if(!panel_open)
+		return NONE
+	if(buildstage != FIRE_ALARM_BUILD_NO_CIRCUIT)
+		return NONE
+
+	loc.balloon_alert_to_viewers("[/obj/item/wallframe/firealarm::name] removed")
+	new /obj/item/wallframe/firealarm(user.drop_location())
+	tool.play_tool_sound(loc)
+	qdel(src)
+	return ITEM_INTERACT_SUCCESS
+
+/obj/machinery/firealarm/wrench_act_secondary(mob/living/user, obj/item/tool)
+	return wrench_act(user, tool)
+
+/obj/machinery/firealarm/tool_act(mob/living/user, obj/item/tool, list/modifiers)
+	. = ..()
+	if(. & ITEM_INTERACT_ANY_BLOCKER)
+		return .
+	if(is_wire_tool(tool))
+		if(panel_open)
+			wires.interact(user)
+			return ITEM_INTERACT_SUCCESS
+
+		balloon_alert(user, "expose the wires first!")
+		return ITEM_INTERACT_BLOCKING
+	return NONE
+
+/obj/machinery/firealarm/item_interaction(mob/living/user, obj/item/tool, list/modifiers)
+	if(!panel_open)
+		return NONE
+
+	if(buildstage == FIRE_ALARM_BUILD_NO_WIRES)
+		if(istype(tool, /obj/item/stack/cable_coil))
+			var/obj/item/stack/cable_coil/coil = tool
+			if(!coil.use(5))
+				balloon_alert(user, "need 5 cables!")
+				return ITEM_INTERACT_BLOCKING
+			buildstage = FIRE_ALARM_BUILD_SECURED
+			balloon_alert(user, "wires installed")
+			update_appearance()
+			return ITEM_INTERACT_SUCCESS
+
+	if(buildstage == FIRE_ALARM_BUILD_NO_CIRCUIT)
+		if(istype(tool, /obj/item/electronics/firealarm))
+			balloon_alert(user, "circuit installed")
+			qdel(tool)
+			buildstage = FIRE_ALARM_BUILD_NO_WIRES
+			update_appearance()
+			return ITEM_INTERACT_SUCCESS
+
+		if(istype(tool, /obj/item/electroadaptive_pseudocircuit))
+			var/obj/item/electroadaptive_pseudocircuit/pseudoc = tool
+			if(!pseudoc.adapt_circuit(user, circuit_cost = 0.015 * STANDARD_CELL_CHARGE))
+				return ITEM_INTERACT_BLOCKING
+			balloon_alert(user, "circuit installed")
+			buildstage = FIRE_ALARM_BUILD_NO_WIRES
+			update_appearance()
+			return ITEM_INTERACT_SUCCESS
```

Comment:
I always find it nicer to split it by items into `[tool]_act(...)` procs, then do the other checks as early returns, but that's not really important and I can see how splitting it by `buildstage` here is more readable

anyhow, any reasons this isn't a switch statement, instead of a series of disconnected ifs?
or well, really it just feels kinda gross to me that a successful `if(buildstage == FIRE_ALARM_BUILD_NO_WIRES)` check can roll into an `if(buildstage == FIRE_ALARM_BUILD_NO_CIRCUIT)` check if the item doesn't match